drivers/timer/nrf_rtc_timer.c

```diff
@@ -1,513 +1,76 @@
-	rtc_compare_set(rtc_past +
-			(max_cycles * sys_clock_hw_cycles_per_tick()));
-
-}
-
-
-/**
- * @brief provides total systicks programmed.
- *
- * returns : total number of sys ticks programmed.
- */
-
-u32_t _get_program_time(void)
-{
-	return expected_sys_ticks;
-}
-
-/**
- * @brief provides total systicks remaining since last programming of RTC.
- *
- * returns : total number of sys ticks remaining since last RTC programming.
- */
-
-u32_t _get_remaining_program_time(void)
-{
-
-	if (!expected_sys_ticks) {
-		return 0;
-	}
-
-	return (expected_sys_ticks - _get_elapsed_program_time());
-}
-
-/**
- * @brief provides total systicks passed since last programming of RTC.
- *
- * returns : total number of sys ticks passed since last RTC programming.
- */
-
-u32_t _get_elapsed_program_time(void)
-{
-	u32_t rtc_elapsed, rtc_past_copy;
+#include <nrf_rtc.h>
+#include <spinlock.h>
 
-	if (!expected_sys_ticks) {
-		return 0;
-	}
-
-	/* Read rtc_past before RTC_COUNTER */
-	rtc_past_copy = rtc_past;
-
-	/* Make sure that compiler will not reverse access to RTC and
-	 * rtc_past.
-	 */
-	compiler_barrier();
-
-	rtc_elapsed = (RTC_COUNTER - rtc_past_copy) & RTC_MASK;
-
-	/* Convert number of Machine cycles to SYS_TICKS */
-	return (rtc_elapsed / sys_clock_hw_cycles_per_tick());
-}
-
-
-/**
- * @brief Sets interrupt for rtc compare value for systick time.
- *
- * This Function does following:-
- * 1. Updates expected_sys_ticks equal to time.
- * 2. Update kernel book keeping for time passed since device bootup.
- * 3. Calls routine to set rtc interrupt.
- */
+#define RTC NRF_RTC1
 
-void _set_time(u32_t time)
-{
-	if (!time) {
-		expected_sys_ticks = 0;
-		return;
-	}
-
-	/* Update expected_sys_ticls to time to programe*/
-	expected_sys_ticks = time;
-	z_tick_set(z_clock_uptime());
-	/* Update rtc_past to track rtc timer count*/
-	rtc_past = (z_tick_get() * sys_clock_hw_cycles_per_tick()) & RTC_MASK;
-
-	expected_sys_ticks = expected_sys_ticks > _get_max_clock_time() ?
-				_get_max_clock_time() : expected_sys_ticks;
-
-	/* Programe RTC compare register to generate interrupt*/
-	rtc_compare_set(rtc_past +
-			(expected_sys_ticks * sys_clock_hw_cycles_per_tick()));
-
-}
-
-/**
- * @brief provides time remaining to reach rtc count overflow.
- *
- * This function returns how many sys RTC remaining for rtc to overflow.
- * This will be required when we will program RTC compare value to maximum
- * possible value.
- *
- * returns : difference between current systick and Maximum possible systick.
- */
-s32_t _get_max_clock_time(void)
-{
-	u32_t rtc_away, sys_away = 0;
+#define COUNTER_MAX 0x00ffffff
+#define CYC_PER_TICK (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC	\
+		      / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+#define MAX_TICKS ((COUNTER_MAX - CYC_PER_TICK) / CYC_PER_TICK)
 
-	/* Calculate time to program into RTC */
-	rtc_away = (RTC_MASK - RTC_COUNTER);
-	rtc_away = rtc_away > RTC_HALF ? RTC_HALF : rtc_away;
+#define MIN_DELAY 32
```

Comment:
Do we really need this minimal delay?
In current implementation, this basically reduces timer resolution, which is a thing we would like to avoid.

Reply:
You have to have a minimum value greater than zero here, otherwise it's possible that the computation of the "next" value based on the "current" timer counter will compute a value that is in the "past" by the timer it arrives in the comparator register.

I picked 1/10th of a tick to be conservative, but given the frequencies here something much lower is probably fine.  You guys are the experts :)